src/fusiontensor/array_cast.jl

```diff
@@ -0,0 +1,193 @@
+# This file defines interface to cast from and to generic array
+
+using BlockArrays: AbstractBlockArray, BlockedArray, blockedrange, blocklengths, findblock
+
+using BlockSparseArrays: BlockSparseArrays, BlockSparseArray
+using GradedUnitRanges: AbstractGradedUnitRange, blocklabels
+using SymmetrySectors: block_dimensions, quantum_dimension
+using TensorAlgebra: contract
+
+# =================================  High level interface  =================================
+
+#### cast from array to symmetric
+function FusionTensor(
+  array::AbstractArray,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+)
+  return cast_from_array(array, codomain_legs, domain_legs)
+end
+
+#### cast from symmetric to array
+function BlockSparseArrays.BlockSparseArray(ft::FusionTensor)
+  return cast_to_array(ft)
+end
+
+# =================================  Low level interface  ==================================
+function cast_from_array(
+  array::AbstractArray,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+)
+  bounds = block_dimensions.((codomain_legs..., domain_legs...))
+  blockarray = BlockedArray(array, bounds...)
+  return cast_from_array(blockarray, codomain_legs, domain_legs)
+end
+
+function cast_from_array(
+  blockarray::AbstractBlockArray,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}};
+  tol::Float64=1e-12,
+)
+  # input validation
+  if length(codomain_legs) + length(domain_legs) != ndims(blockarray)  # compile time
+    throw(DomainError("legs are incompatible with array ndims"))
+  end
+  if quantum_dimension.((codomain_legs..., domain_legs...)) != size(blockarray)
+    throw(DomainError("legs dimensions are incompatible with array"))
+  end
+
+  ft = unsafe_cast_from_array(blockarray, codomain_legs, domain_legs)
+
+  # if blockarray is not G-invariant, norm(ft) < norm(blockarray)
+  if abs(norm(ft) - norm(blockarray)) > tol
+    throw(
+      InexactError(
+        :FusionTensor, typeof(blockarray), typeof(codomain_legs), typeof(domain_legs)
+      ),
+    )
+  end
+  return ft
+end
+
+function unsafe_cast_from_array(
```

Comment:
I'm not a fan of using the term "unsafe" since it is a bit overloaded and underdefined. I would prefer to make it more explicit in the name of the function what isn't being checked/what is being assumed about the input.